logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java

```diff
@@ -271,6 +321,44 @@ public void setCurrentReaderAndPosition(Path segmentPath, long position) throws
                 .ifPresent(reader -> currentReader = reader);
     }
 
+    private void removeSegmentsBefore(Path validSegment) throws IOException {
+        final Comparator<Path> fileTimeAndName = ((Comparator<Path>) this::byFileFimestamp)
+                .thenComparingInt(DeadLetterQueueUtils::extractSegmentId);
+
+        Files.list(queuePath)
+                .filter(p -> fileTimeAndName.compare(p, validSegment) < 0)
+                .forEach(this::trashSegment);
+    }
+
+    private int byFileFimestamp(Path p1, Path p2) {
```

Comment:
> `byFileFimestamp` seems has typo

yes it is

> avoid attaching numbers to vars like `p1`, `p2`

This is a common practice in [JDK](https://github.com/openjdk/jdk17/blob/74007890bb9a3fa3a65683a3f480e399f2b1a0b6/src/java.base/share/classes/java/util/Comparator.java#L440). I don't think we have conventions on it. While I would agree in general with your concern, in this case I think that `firsthPath` is not such different from `p1`, needs more attention in reading to understand what's is intended for. `p1` and `p2` immediately says that they are the same type of object and have no ordering semantic,  the number is just a postfix.
I would prefer to keep them.

> if IO exception happens when getting last modified timestamp, would it be a right approach if we consider both files are equal?

If we get an IO error in any of two, we simply can't compare and the idea is to move forward the decision to the next comparator, which leverages the naming convention. In case we reach `MAX_INT`, so that is ends in `MAX_INT.log` being older than `0.log`, we should be very unlucky to get an IO exception while comparing by timestamp. 
A way to solve this is to put also the timestamp, other than the sequence segment number, in the filename.